### ISSUE TITLE
Graph Collections in the C API

### ISSUE THREAD

**Issue by eaplatanios** [tensorflow/tensorflow#8962]
Hi,

In Python we have graph collections (e.g., TRAINABLE_VARIABLES) which are stored in the Graph class. I assume that these are somehow serialized in the graph protobuf so that when a GraphDef is imported, the relevant collections are imported too. However, I do not see any option in the C API for defining such collections? Is there something that I am missing? If not, why not include that functionality in the C API?

Thank you,
Anthony

**Comment by asimshankar** [comment]
You are correct in that graph collections in the Graph class in Python (which make it to the `MetaGraphDef` protocol buffer, which in turn is exported in the saved model format (`TF_LoadSessionFromSavedModel`)) are not part of the C API - users will have to write their own collection management on top of the existing primitives.

Contributions are welcome (though perhaps want to discuss design a bit before pursuing it)

**Comment by eaplatanios** [comment]
I see. I'm currently developing a Scala API and that's what I was thinking of doing (i.e., independent management of collections). However, I was wondering what would be the right way to serialize the graph so that it can also be imported in the Python API and contain the same information. You are implying I should be using the MetaGraphDef protocol buffer and the saved model format used in the Python API, right?

In terms of design, wouldn't it make more sense to make collections part of the GraphDef and add functions for manipulating them in the C API?

**Comment by asimshankar** [comment]
I'm guessing you mean making functions part of the `TF_Graph` and add functions for manipulating them? Yes, that sounds reasonable, just not something we've gotten around to yet and hence contributions welcome :)

(BTW, just want to make sure you're aware of the [TensorFlow Java API](https://www.tensorflow.org/api_docs/java/reference/org/tensorflow/package-summary) which may be useful for Scala as well)

**Comment by eaplatanios** [comment]
Ok for now I'll use a temporary Scala-only solution, but will also look into what would be required for supporting that in the C API.

I am aware of Java API. I actually took the JNI bindings from that project, extended them to support more features of the C API (e.g., control inputs, obtaining all ops in the graph, among others), and built an entirely new Scala library on top of it. It is loosely based on the Java library but supports more features (such as a createWith function for specifying name scopes, control dependencies, device contexts, among others). My goal is to make it support much of what the Python API supports so I am currently going through the Python code and adding features. I plan to open it to the public once it is complete enough to be usable in wider contexts than the Java API.

**Comment by suiyuan2009** [comment]
I'll work on this!
